arangod/RestHandler/RestAdminClusterHandler.cpp

```diff
@@ -0,0 +1,1563 @@
+
+  VPackSlice dbServers = plan.get(planPath->dBServers()->vec());
+  bool serversFound = dbServers.isObject() && dbServers.hasKey(ctx->fromServer) && dbServers.hasKey(ctx->toServer);
+  if (!serversFound) {
+    generateError(ResponseCode::NOT_FOUND, TRI_ERROR_HTTP_NOT_FOUND,
+      "one or both dbservers not found");
+    return futures::makeFuture();
+  }
+
+  VPackSlice collection = plan.get(planPath->collections()
+    ->database(ctx->database)->collection(ctx->collectionID)->vec());
+  if (!collection.isObject()) {
+    generateError(ResponseCode::NOT_FOUND, TRI_ERROR_HTTP_NOT_FOUND,
+      "database/collection not found");
+    return futures::makeFuture();
+  }
+
+  if (collection.hasKey("distributeShardsLike")) {
+    generateError(ResponseCode::FORBIDDEN, TRI_ERROR_HTTP_FORBIDDEN,
+      "MoveShard only allowed for collections which have distributeShardsLike unset.");
+    return futures::makeFuture();
+  }
+
+  VPackSlice shard = collection.get(std::vector<std::string>{"shards", ctx->shard});
+  if (!shard.isArray()) {
+    generateError(ResponseCode::NOT_FOUND, TRI_ERROR_HTTP_NOT_FOUND,
+      "shard not found");
+    return futures::makeFuture();
+  }
+
+  bool fromFound = false;
+  bool isLeader = false;
+  for(VPackArrayIterator i(shard); i != i.end(); i++) {
+    if (i.value().isEqualString(ctx->fromServer)) {
+      isLeader = i.isFirst();
+      fromFound = true;
+    }
+  }
+
+  if (!fromFound) {
+    generateError(ResponseCode::NOT_FOUND, TRI_ERROR_HTTP_NOT_FOUND,
+      "shard is not located on the server");
+    return futures::makeFuture();
+  }
+
+  std::string jobId = std::to_string(server().getFeature<ClusterFeature>().clusterInfo().uniqid());
+  auto jobToDoPath = arangodb::cluster::paths::root()->arango()->target()->toDo()->job(jobId);
+
+  VPackBuffer<uint8_t> trx;
+  {
+    VPackBuilder builder(trx);
+    arangodb::agency::envelope<VPackBuilder>::create(builder)
+      .write()
+        .emplace(jobToDoPath->str(), [&](VPackBuilder &builder) {
+          builder.add("type", VPackValue("moveShard"));
+          builder.add("database", VPackValue(ctx->database));
+          builder.add("collection", collection.get("id"));
+          builder.add("jobId", VPackValue(jobId));
+          builder.add("shard", VPackValue(ctx->shard));
+          builder.add("fromServer", VPackValue(ctx->fromServer));
+          builder.add("toServer", VPackValue(ctx->toServer));
+          builder.add("isLeader", VPackValue(isLeader));
+          builder.add("remainsFollower", VPackValue(isLeader));
+          builder.add("creator", VPackValue(ServerState::instance()->getId()));
+          builder.add("timeCreated", VPackValue(timepointToString(std::chrono::system_clock::now())));
+        })
+        .done()
+      .done();
+  }
+
+  auto self(shared_from_this());
+
+  return AsyncAgencyComm().sendWriteTransaction(20s, std::move(trx)).thenValue(
+    [this, ctx = std::move(ctx), jobId = std::move(jobId)](AsyncAgencyCommResult &&result) {
```

Comment:
```suggestion
    [this, ctx = std::move(ctx), jobId = std::move(jobId)](AsyncAgencyCommResult&& result) {
```
do we need to capture `self` as well?

Reply:
This lambda is executed (or destroyed) strictly before the `thenFinal` lambda in `waitForFuture` which captures `self`.